### PR TITLE
Make it possible to use a wrapped EventLoop with a Channel

### PR DESCRIPTION
**Pull request by normanmaurer**
Motiviation:

Because of how we implemented the registration / deregistration of an EventLoop it was not possible to wrap an EventLoop implementation and use it with a Channel.

Modification:

- Introduce EventLoop.Unsafe which is responsible for the actual registration.
- Move validation of EventLoop / Channel combo to the EventLoop
- Add unit test that verifies that wrapping works

Result:

Be able to wrap an EventLoop and so add some extra functionality.

### CODE REVIEW

**Comment by normanmaurer** [comment]
This depends on https://github.com/netty/netty/pull/8587 ...

 I will open another PR on top of this one here shortly, which should complete the EventLoop rework and allow easy customisation etc.

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by normanmaurer** [comment]
@stef1927 can you have a look here as well as this is the basis for 
https://github.com/netty/netty/pull/8680

**Comment by normanmaurer** [comment]
Also ping @carl-mastrangelo @ejona86

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollRegistration.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel.epoll;
+
+import io.netty.channel.unix.IovArray;
+
+import java.io.IOException;
+
+interface EpollRegistration {
```
Comment (carl-mastrangelo) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollRegistration.java]: These would probably benefit from some docs.

File: transport-native-kqueue/src/main/java/io/netty/channel/kqueue/KQueueRegistration.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel.kqueue;
+
+import io.netty.channel.unix.IovArray;
+
+interface KQueueRegistration {
```
Comment (carl-mastrangelo) [inline transport-native-kqueue/src/main/java/io/netty/channel/kqueue/KQueueRegistration.java]: same comment.

File: transport/src/main/java/io/netty/channel/local/LocalChannel.java [hunk 3]
```diff
@@ -506,5 +456,48 @@ public void connect(final SocketAddress remoteAddress,
...
+            // deregistered / registered later again.
+            //
+            // See https://github.com/netty/netty/issues/2400
+            if (peer != null && parent() != null) {
+                // Store the peer in a local variable as it may be set to null if doClose() is called.
+                // See https://github.com/netty/netty/issues/2144
+                final LocalChannel peer = LocalChannel.this.peer;
+                state = State.CONNECTED;
+
+                peer.remoteAddress = parent() == null ? null : parent().localAddress();
```
Comment (carl-mastrangelo) [inline transport/src/main/java/io/netty/channel/local/LocalChannel.java]: unless parent can change after line 467, I think this conditional is not needed.

File: transport/src/main/java/io/netty/channel/local/LocalEventLoop.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+import io.netty.channel.Channel;
+import io.netty.channel.EventLoopGroup;
+import io.netty.channel.SingleThreadEventLoop;
+import io.netty.util.concurrent.DefaultThreadFactory;
+import io.netty.util.internal.StringUtil;
+
+import java.util.concurrent.Executor;
+import java.util.concurrent.ThreadFactory;
+
+public class LocalEventLoop extends SingleThreadEventLoop {
```
Comment (carl-mastrangelo) [inline transport/src/main/java/io/netty/channel/local/LocalEventLoop.java]: This probably needs a javadoc comment as well.

File: transport/src/main/java/io/netty/channel/local/LocalEventLoopGroup.java [hunk 5]
```diff
@@ -48,4 +51,19 @@ public LocalEventLoopGroup(int nThreads) {
...
+     *
+     * @param nThreads          the number of threads to use
+     * @param executor          the Executor to use, or {@code null} if the default should be used.
+     */
+    public LocalEventLoopGroup(int nThreads, Executor executor) {
+        super(nThreads, executor);
+    }
+
+    @Override
+    protected EventLoop newChild(Executor executor, Object... args) throws Exception {
```
Comment (carl-mastrangelo) [inline transport/src/main/java/io/netty/channel/local/LocalEventLoopGroup.java]: Should these args ever be set?  Might be worth failing if any are.

File: transport/src/test/java/io/netty/channel/AbstractChannelTest.java [hunk 6]
```diff
@@ -48,6 +58,17 @@ public void ensureSubsequentRegistrationDoesNotFireActive() throws Throwable {
         final EventLoop eventLoop = mock(EventLoop.class);
         // This allows us to have a single-threaded test
         when(eventLoop.inEventLoop()).thenReturn(true);
+        when(eventLoop.unsafe()).thenReturn(new EventLoop.Unsafe() {
```
Comment (carl-mastrangelo) [inline transport/src/test/java/io/netty/channel/AbstractChannelTest.java]: I think you can return `mock(EventLoop.Unsafe.class)`

File: transport/src/test/java/io/netty/channel/SingleThreadEventLoopTest.java [hunk 7]
```diff
@@ -484,6 +484,11 @@ protected void run() {
         protected void cleanup() {
             cleanedUp.incrementAndGet();
         }
+
+        @Override
+        public Unsafe unsafe() {
+            return null;
```
Comment (carl-mastrangelo) [inline transport/src/test/java/io/netty/channel/SingleThreadEventLoopTest.java]: should this ever be called?    Maybe throw?

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollRegistration.java (hunk 1 above)
Comment (normanmaurer) [reply transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollRegistration.java]: sure I can do... I did not bother as its package private anyway.

File: transport-native-kqueue/src/main/java/io/netty/channel/kqueue/KQueueRegistration.java (hunk 2 above)
Comment (normanmaurer) [reply transport-native-kqueue/src/main/java/io/netty/channel/kqueue/KQueueRegistration.java]: sure I can do... I did not bother as its package private anyway.

File: transport/src/main/java/io/netty/channel/local/LocalChannel.java (hunk 3 above)
Comment (normanmaurer) [reply transport/src/main/java/io/netty/channel/local/LocalChannel.java]: The code was just moved... So I would prefer to not adjust it here and do a followup pr if needed

File: transport/src/main/java/io/netty/channel/local/LocalEventLoop.java (hunk 4 above)
Comment (normanmaurer) [reply transport/src/main/java/io/netty/channel/local/LocalEventLoop.java]: I made this package-private as the user should never create an instance directly but use `LocalEventLoopGroup`

**Comment by stef1927** [comment]
> @stef1927 can you have a look here as well as this is the basis for #8680

It looks good to me, thanks for the ping.

**Comment by normanmaurer** [comment]
@stef1927 please "approve" :)

**Comment by normanmaurer** [comment]
Let me merge this one to be able to make progress... @carl-mastrangelo please open a PR as followup if there is other stuff you want to have changed.

**Pull request merged**